File: src/main/java/ru/citeck/ecos/predicate/model/OrPredicate.java
package ru.citeck.ecos.predicate.model;

import com.fasterxml.jackson.annotation.JsonProperty;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;

public class OrPredicate extends ComposedPredicate {

    private static final String TYPE = "or";

    @JsonProperty("t")
    String getType() {
        return TYPE;
    }

    public static List<String> getTypes() {
        return Collections.singletonList(TYPE);
    }

    public static OrPredicate of(Predicate... predicates) {
        OrPredicate or = new OrPredicate();
        or.setPredicates(Arrays.asList(predicates));
        return or;
    }

    @Override
    public String toString() {
        return "(" + getPredicates().stream()
                                    .map(Object::toString)
                                    .collect(Collectors.joining(" OR ")) + ")";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        return super.equals(o);
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records2/predicate/model/OrPredicate.kt
package ru.citeck.ecos.records2.predicate.model

import ecos.com.fasterxml.jackson210.annotation.JsonProperty

class OrPredicate : ComposedPredicate() {

    companion object {

        @JsonProperty("t")
        const val TYPE = "or"

        @JvmStatic
        fun getTypes(): List<String> {
            return listOf(TYPE)
        }

        @JvmStatic
        fun of(vararg predicates: Predicate): OrPredicate {
            return of(listOf(*predicates))
        }

        @JvmStatic
        fun of(predicates: List<Predicate>): OrPredicate {
            val or = OrPredicate()
            or.setPredicates(predicates)
            return or
        }
    }

    @JsonProperty("t")
    fun getType(): String {
        return TYPE
    }

    override fun <T : Predicate> copy(): T {
        val copy = OrPredicate()
        copy.setPredicates(getPredicates().map { it.copy<Predicate>() })
        @Suppress("UNCHECKED_CAST")
        return copy as T
    }

    override fun toString(): String {
        return "{\"t\":\"$TYPE\",\"val\":[${getPredicates().joinToString(",") { it.toString() }}]}"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || OrPredicate::class.java != other::class.java) {
            return false
        }
        other as OrPredicate
        return getPredicates() == other.getPredicates()
    }

    override fun hashCode(): Int {
        return getPredicates().hashCode()
    }
}
